### PR TITLE
[RISCV] Pass subvector type to isLegalInterleavedAccessType in getInterleavedMemoryOpCost.

### PR DESCRIPTION
**Pull request by topperc**
isLegalInterleavedAccessType expects the subvector type, but getInterleavedMemoryOpCost is called with the full vector type. So we need to divide by Factor.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -612,14 +612,19 @@ InstructionCost RISCVTTIImpl::getInterleavedMemoryOpCost(
...
-      // Should the memory op cost of these be cheaper?
-      if (TLI->isLegalInterleavedAccessType(LegalVTy, Factor, Alignment,
+      auto *SubVecTy =
+          VectorType::get(VTy->getElementType(),
+                          VTy->getElementCount().divideCoefficientBy(Factor));
+
+      if (VTy->getElementCount().isKnownMultipleOf(Factor) &&
+          TLI->isLegalInterleavedAccessType(SubVecTy, Factor, Alignment,
                                             AddressSpace, DL)) {
+        // FIXME: We use the memory op cost of the *legalized* type here,u
```
Comment (arcbbb) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:622]: the trailing u can be removed.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -612,14 +612,19 @@ InstructionCost RISCVTTIImpl::getInterleavedMemoryOpCost(
...
+        // because it's getMemoryOpCost returns a really expensive cost for
+        // types like <6 x i8>, which show up when doing interleaves of
+        // Factor=3i etc. Should the memory op cost of these be cheaper?
```
Comment (arcbbb) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:625]: Factor=3i -> 3.

**Review by arcbbb — APPROVED** [review]
LGTM.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -612,14 +612,19 @@ InstructionCost RISCVTTIImpl::getInterleavedMemoryOpCost(
...
-      // FIXME: We use the memory op cost of the *legalized* type here, becuase
-      // it's getMemoryOpCost returns a really expensive cost for types like
-      // <6 x i8>, which show up when doing interleaves of Factor=3 etc.
-      // Should the memory op cost of these be cheaper?
-      if (TLI->isLegalInterleavedAccessType(LegalVTy, Factor, Alignment,
+      auto *SubVecTy =
+          VectorType::get(VTy->getElementType(),
+                          VTy->getElementCount().divideCoefficientBy(Factor));
+
+      if (VTy->getElementCount().isKnownMultipleOf(Factor) &&
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:619]: Is getInterleavedMemoryOpCost ever called with a VTy that isn't a multiple of Factor?

**Review by lukel97 — APPROVED** [review]

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 3 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:619]: Not sure. AArch64 checked divisibility so I did too.

**Review by Mel-Chen — APPROVED** [review]
LG

**Pull request merged**